### ISSUE TITLE
Questions about the protocol design of pull-streams

### ISSUE THREAD

**Issue by elavoie** [pull-stream/pull-stream#100]
I am looking deeper into the aborting and error handling parts of the pull-stream protocol and some things seem fuzzy to me and would need clarifications. But first I would need validation of my understanding of the protocol. Let me explain it in three steps, each capturing progressively more behaviours then the previous. I will explain my questions along the way.

# (1) Base: Demand-driven stream processing

All communication between modules is initiated by the module downstream (consumer) and propagates upstream towards the producer. So:

The consumer first *asks* for a value. Then the producer has two choices, it can either answer with a *value* or say the stream is *done*.

Nothing can go wrong in that model. The consumer asks for a value, then receive it, then asks for another one, etc. until the producer does not have any more values and declares the stream done.

The *done* answer serves as a synchronization point. When a module receives *done*, it knows the module upstream has finished cleaning up after itself in addition to terminating the stream and no more values will ever be produced.

Possibilities:

ask->value: Normal case, receive a value
ask->done: Normal case, stream finished

** Question: Can the consumer continue asking for values after receiving a done? **

If so, when does the producer know that it won't receive anymore requests?

# (2) Base + Early Aborting

Now, maybe the consumer does not need all the values of the producer. Therefore, it needs to be able to tell the producer to stop even though it might still have more values to produce.

The consumer may therefore *abort* in addition to *ask* for a value. The *ask* protocol stays the same. After an abort, the producer has only one choice, which is to say that it is *done* aborting. 

The *done* answer serves the same purpose as for the base case and does it also for the aborting case.

Possibilities:

ask->value: Normal case, receive a value
ask->done: Normal case, stream finished
abort->done: Consumer early abort, producer closed

Now the early aborting introduces a tricky case, the consumer may abort before receiving a value, which gives two different sequences of events.

ask1, abort2, value1, done2
ask1, abort2, done2, value1

The second case is problematic because suddenly the done event does not act as a synchronization point anymore: callbacks may be called after the module upstream has said it was done.

**Question: Is that allowed in the pull-stream protocol?**

An easy way to prevent that problem would be to prevent the consumer to abort before it received a value, therefore the only valid sequence would be:

ask1, value1, abort2, done2

and the done event would again become a synchronization point.

**Question: Can a consumer abort multiple times?**

If so, then the stream is never terminated as there could always be another abort in the future. It makes it harder to reason about...

**Question: Can the producer answer an abort with done without ever returning a value for the previous ask?**

Now if so that is rather tricky the consumer may never know if the done event means:

ask1, abort2, done2 (terminated)

or 

ask1, abort2, done2 .... (arbitrarily large delay) value1

# (3) Base + Early Aborting + Errors

Now what if something can go wrong in modules, such that a communication timeout? The protocol is extended with error signalling in both directions. Therefore the consumer may signal an *error* to terminate the processing. In answer to an *ask*, an *abort*, or an *error*, the producer may also answer with an *error*. That therefore gives the following possibilities:

ask->value: Normal case, receive a value
ask->done: Normal case, stream finished
ask->error: Consumer ok, producer failed
abort->done: Consumer early abort, producer closed
abort->error: Consumer early abort, producer failed
error->done: Consumer failed, producer closed
error->error: Consumer failed, producer failed too

Now in addition to these 8 cases, there are additional tricky edge cases with more combinations. If a module may abort multiple times, that makes error handling more difficult, that allows things like a module failing after *a second abort*. 

**Question: Does the protocol allows a failure after multiple aborts?**

**Question: Should failures be propagated through the pipeline?**

There is another tricky point. An error first happens in a specific module but is then communicated to another module, either upstream or downstream.

If the failure is propagated, then we loose *locality*: receiving an error does not mean the module immediately before or after us failed, it means some module somewhere failed. Moreover, what if the propagation of an error *introduces other errors* in other modules? What error should be propagated?

In addition, it may be hard to retrace where the error happened in the first place for debugging purposes. The other option would be to keep locality by not propagating an error but rather aborting. However, doing so will silently suppress errors.

It would seem that a better option would be to make error reporting *orthogonal* to the stream: when a module fails, it adds an error on its error stream, then aborts (or the opposite, I don't know which one is better). The stream itself could be a pull-stream from model (2), where each module is also an error producer. That way the error of a module may still be tied to that module but would not propagate through the main stream. When a module finishes it can (or must?) close its error stream.

Moreover, the errors may be listened to only if required. When building a stream from modules that cannot fail (or the failure handling does not need a separate behaviour from aborting or normal completion), one can program strictly in model (2).

What do you guys think?

**Comment by nichoth** [comment]
1. *Can the consumer continue asking for values after receiving a done?*
No, that would be an incorrect sink

2. I'm not sure I follow the second question. The stream semantics are basically request/response. The request can be 'abort', in which case the source stream will send a 'done' or 'error' response.

3. *Can a consumer abort multiple times?*
no

4. *Can the producer answer an abort with done without ever returning a value for the previous ask?*
No. The sink should not abort before recieving a response to it's request for a value.

5. *Does the protocol allows a failure after multiple aborts?*
No, see above. But the source can return an error in response to an abort request.

6. *Should failures be propagated through the pipeline?*
Yes errors are simply passed through

It can be tricky to implement everything correctly, so generally you want to just compose exiting streams whenever possible instead of writing a new one.

**Comment by elavoie** [comment]
> I'm not sure I follow the second question. The stream semantics are basically request/response. The request can be 'abort', in which case the source stream will send a 'done' or 'error' response.

Suppose a sink first 'ask' (````read(false, cb1)````), then 'aborts' before 'cb1' was called (````read(true, cb2)````). Is it possible that 'cb2' will be called before 'cb1'?

> It can be tricky to implement everything correctly, so generally you want to just compose existing streams whenever possible instead of writing a new one.

Well I know, but you can't escape it when you are writing new modules. I wrote [these](https://github.com/elavoie/pull-lend-stream) [two](https://github.com/elavoie/pull-lend) one year ago and I am still not convinced they are fully correct. They are rather tricky as they are arbitraging between multiple concurrent sub-streams. I am encountering all these corner cases while trying to make the spec for the modules complete, i.e. covering all possible cases.

**Comment by elavoie** [comment]
Actually, real modules such as [count](https://github.com/pull-stream/pull-stream/blob/master/sources/count.js) support an additional request from the consumer:

*done* (*abort* without a callback, therefore not expecting an answer)

It would be really useful to have all allowed and forbidden interaction patterns documented. The callback implementation in JavaScript allows sequences of events that do are not part of expected interaction patterns. Where should I write such a document?

**Comment by elavoie** [comment]
@nichoth 
> About errors, in practice I find it's not too much of a problem, and much clearer compared to say, node streams. One pattern I like is to pipe a source stream through another stream that handles the error. This is useful if you are combining multiple source streams, for example, and want more control over how the error affects the pipeline. See the pull-catch module and pull-many.

So in effect you are pulling out the error from the main stream and dealing with it in a callback. That would reinforce my suggestion that errors should be orthogonal to the main stream and originate from the individual modules. Maybe pull-catch is the simplest way to do that.

How often do you use the second test case of pull-catch (propagate the error) or the third (modify what is propagated)?

Moreover, your test cases handle errors coming from upstream but I don't see any test case where errors come from downstream. How is that handled?

**Comment by nichoth** [comment]
> Suppose a sink first 'ask' (read(false, cb1)), then 'aborts' before 'cb1' was called (read(true, cb2)). Is it possible that 'cb2' will be called before 'cb1'?

I think it is incorrect for the sink to abort before getting a response to the first callback. There should be only 1 request pending at any given time.

> ctually, real modules such as count support an additional request from the consumer:
> done (read(true) or abort without a callback, therefore not expecting an answer)

So here count is a source stream. A source can send an 'end' signal like `cb(true)`, and the sink should not make any more requests after that, so you don't need to pass any other parameters.

>Moreover, your test cases handle errors coming from upstream but I don't see any test case where errors come from downstream. How is that handled?

If the sink sends 'abort', it is simply passed on to the source streams. If the sink dies, that means it is impossible to read any more values. I haven't seen a need to do anything else in that situation. 

> How often do you use the second test case of pull-catch (propagate the error) or the third (modify what is propagated)?

Conceptually it is like a map. If one source errors in a multiplexed stream, we may want to send a message to the sink telling it about the error, but keep the other sources alive. So you could for example show an error message in a UI. To be honest I haven't used pull-catch all that much in real life, I just included multiple scenarios for completeness. 

This is fun. I feel like we're taking the stream mid term.

**Comment by nichoth** [comment]
> Moreover, your test cases handle errors coming from upstream but I don't see any test case where errors come from downstream. How is that handled?

Actually I do remember making a through stream `on-abort`, that would listen for an abort signal and call a cb. I used it for one thing, but I feel like it is not necessary.

**Comment by elavoie** [comment]
>> Suppose a sink first 'ask' (read(false, cb1)), then 'aborts' before 'cb1' was called (read(true, cb2)). Is it possible that 'cb2' will be called before 'cb1'?

>I think it is incorrect for the sink to abort before getting a response to the first callback. There should be only 1 request pending at any given time.

From the [pull-stream README](https://github.com/pull-stream/pull-stream#sink-abort): 

> Sinks do not normally end the stream, but if they decide they do not need any more data they may "abort" the source by calling read(true, cb). A abort (read(true, cb)) may be called before a preceding read call has called back.

That seems to contradict your understanding. I think it means we need to make things clearer ;-).

----------------------------------------------------------------------------------------------------

>> Actually, real modules such as count support an additional request from the consumer:
done (read(true) or abort without a callback, therefore not expecting an answer)

> So here count is a source stream. A source can send an 'end' signal like cb(true), and the sink should not make any more requests after that, so you don't need to pass any other parameters.

You are describing the ````ask->done```` case which is different from what I was describing. What I noticed from the implementation of pull-count is that it anticipates a case where it might be called without a callback. I called that the ````done```` signal, which contrary to all the other cases cannot receive an answer because there is no callback associated. That is an exception to the request/response structure of the protocol.

After thinking some more about it, it seems the done signal coming from the sink is the only way from the producer to know the processing for the entire pipeline has finished. This is because the sink initiates all communications. So even when the source responds with ````cb(true)```` saying I  am done, it cannot know when the answer has traversed the entire pipeline. On the other hand, when the sink sends the done signal, it knows it will never send any more requests and no answer will ever come back because there are no callbacks. Just as requests are initiated from the sink, so should the synchronization to signal the end of processing.

Actually, I re-read the pull-stream [README.md](https://github.com/pull-stream/pull-stream#handling-endaborterror-in-through-streams), it mentions:

> Rules for implementing read in a through stream:
> 1. Sink wants to stop. sink aborts the through
>    just forward the exact read() call to your source, any *future read calls* should cb(true).

That suggests ````abort->done```` may happen multiple times in the future. So I would suggest to add the rule that done initiated from the sink (````read(true)````) should happen only once and be the last event of the sink. That way, when the source receives it, it knows all processing is over. 

----------------------------------------------------------------------------------------------------

> This is fun. I feel like we're taking the stream mid term.

I really like the pull-stream protocol, it is simple and elegant. Moreover it grew a healthy community and seemed to have evolved organically to cover many use cases. Now I am trying to have a clear spec of the complete protocol to reason about correctness. Indeed corner cases study do sound like mid-terms exams in protocol design ;-). I hope this one is more fun and much less stressful than most university exams!

**Comment by nichoth** [comment]
> Sinks do not normally end the stream, but if they decide they do not need any more data they may "abort" the source by calling read(true, cb). A abort (read(true, cb)) may be called before a preceding read call has called back.

Ah ha, yes, it is a little more complicated than my mental model. I wonder what the reason is for this choice, being able to abort before getting a response. I'm guessing it is to save some resources in some cases, since the source could possibly stop what it is doing early.

> anticipates a case where it might be called without a callback

oh yeah, there is a check to see if the cb exists: ` if(end) return cb && cb(end)`. I was under the impression that a sink should *always* call the source with a cb.

**Comment by nichoth** [comment]
> it seems the done signal coming from the sink is the only way from the producer to know the processing for the entire pipeline has finished

To me, it seems like a source should not need to know if the entire pipeline is done. The source is able either end on it's own, in which case it can free whatever resources it is using, or it gets the 'abort' signal and ends. I can't think of a reason why it would need to know any more details about what is happening downstream.

> Just as requests are initiated from the sink, so should the synchronization to signal the end of processing. 

Does this mean that if a source ends, it would send `cb(true)`, then should be called again by the sink with `read(true)`, which signals to the source that the sink is all done?

**Comment by elavoie** [comment]
> Does this mean that if a source ends, it would send cb(true), then should be called again by the sink with read(true), which signals to the source that the sink is all done?

Yes that is what I am suggesting. If we are to support multiple aborts by the sink as [the wording in the README suggests](https://github.com/pull-stream/pull-stream#handling-endaborterror-in-through-streams), I think that it is the only way to know in any modules upstream of the sink that no more events will happen.

**Comment by regular** [comment]
>  I wonder what the reason is for this choice, being able to abort before getting a response.

@nichoth isn't that the only way you can abort an async source that is not producing any value (anymore)? Like in a network stream where the other end simply doesn't send any data. You still want to be able to hang up, even if the phone line is silent.

**Comment by elavoie** [comment]
@regular that makes sense.  In that case, should the async source be forbidden from returning a value *after it answered done to the abort*? Is there a case where it would make sense for a value to be returned?

**Comment by nichoth** [comment]
@regular yes that makes sense

@elavoie i would think you would not want to return any value after 'done' since it seems to not make sense semantically

**Comment by elavoie** [comment]
After thinking some more about it, I think I should refrain from any proposition to extend or change the protocol until I actually clearly described all the protocol patterns currently used by existing pull-stream modules. That way, it should make it easier to see which backward-compatible extensions are possible and which non-backward compatible changes could possibly be needed. I think the current callback protocol is both rich enough to accommodate many possibilities while supporting the most common cases easily which may explain its success.

**Comment by nichoth** [comment]
Thanks for starting this discussion. It's nice to learn about some of the nuances we talked about

**Comment by dominictarr** [comment]
@elavoie @nichoth @regular yes - the abort may be called before the previous ask has called back. This is necessary to abort live streams that have paused. you can do a lot with pull-streams before needing this (actually I didn't realize this was the case until I started working on ssb, but could add it with few changes)

in an out of turn abort, the callback to the previous ask should be called with the end, then the callback to the abort should be called.

There are only a couple of cases.
``` 
//valid order for IN turn abort
while(zero_or_more) {
  read(null, cb1)
  cb1(null, data)
}
if(normal_end) {
  read(null, cb2)
  cb2(true)
}
else if(in_turn_abort) {
  read(true, cb2) //in turn abort
  cb2(true)
}
else (out_of_turn_abort) {
  read(null, cb2)
  read(true, cb3) //out of turn abort, before cb1 has been called.
  cb2(true)
  cb3(true)
}
```
note, the above also allows not reading any data, just aborting immediately.

**Comment by dominictarr** [comment]
if you abort a stream out of turn, but the implementation is still halfway through getting say the next 3 values (say, reading a block from a file) and then the abort comes in (causing the file to close) the implementation should just throw out the data values that were asked for and cb "done" to the ask and the abort.

**Comment by elavoie** [comment]
Do modules support *asking multiple values* before processing the answers?

Example: ````read(false, cb1); read(false, cb2);````

That would be equivalent to the previous case where the second read is an abort. In that case, it would seem the minimal constraint should be that cb1 be called before cb2. Moreover, if the source is done for cb1, any other following callbacks should be done too.

Note: I [created a repository](https://github.com/pull-stream/specification) to document in a clear and precise way the discussions on this thread so we can refer to it in the future.

**Comment by dominictarr** [comment]
@elavoie no. I considered that in the early days, but that makes it really complicated.
in cases where you want to say, ensure that data is flowing in really fast, or that there is some amount buffered use a module that provides that like: https://www.npmjs.com/package/pull-high-watermark

**Comment by elavoie** [comment]
@dominictarr So if I understand well then, imagine I have a pull-stream with a source Src, a sink Sk, and two through (transformers) T1 T2. Imagine Src produces values instantly and Sk consumes them instantly but the two transformers each take a second to process a value each using a different hardware core. Since each request is issued only after the previous answer has been received, then there could be no parallelism in this basic case: the throughput of the pipeline would be 1/2 values per second. However, if the stream was producer-driven we could reach 1 value per second even though each value requires 2s of processing time. Am I right?

In a sense, the main aim of the pull-stream protocol is to ease composition of processing functions and not use a stream paradigm to improve throughput.

**Comment by nichoth** [comment]
@elavoie does pull-paramap solve the performance optimization in this example? https://www.npmjs.com/package/pull-paramap

**Comment by elavoie** [comment]
Actually if I understand paramap correctly, it is a different kind of parallelism: *the same operation in the same through happens in parallel on multiple values*. The question I am asking is whether *different operations happening in different throughs* may process values in parallel.

An analogy would be, imagine you have a conveyor belt with figurines to paint. Using paramap you may ask multiple people to paint the face in parallel and put back the figurines on the conveyor belt in the right order when they are done. In my example, I am asking whether one person may paint the face and then pass the figuring to a second one who would paint the body, while the first person may paint the face of the next one right away. I think the synchronization of the sink currently asks for a new figuring to be painted, then the first person paints the face, pass it to the second person who paints the body then pass it to the sink. The sink then asks for the next figurine. Since the sink waits for the figuring to be completely painted you obtain half the throughput that could possibly achieved if both person worked in parallel.

**Comment by nichoth** [comment]
What if your source data included something like different data types, then in the paramap through you do different operations depending on the data type?

so like

```js
paramap((value, cb) => asyncOperation[value.type](value.data, cb))
```

effectively multiplexing the source events

So we have multiple different operations happening in parallel, and if you connect another through stream, it would get values as soon as they are available from paramap (not necessarily in the same order).

---------------------------------------

actually, I think what I wrote above may be a different situation. Sometimes I will return a stream from a pull.map, maybe this is closer to the question.

So you have something like

```js
S( source,
    S.map( value => pullCat([ value, asyncSource(value) ]) ),
    pullFlatMerge(),  // this "flattens" the nested stream, without preserving event order
    S.map(/* this through uses the original value, then also gets the async value from above */)
    sink
)
```

**Comment by nichoth** [comment]
hmm, so actually, the second transform depends on the results from the first async operation, which is not handled, but if I've understood, a new value is only pulled from the source once all the throughs have finished. ok

**Comment by elavoie** [comment]
@nichoth I am not sure I understand how your solution solves the problem. 

What I can think of to mitigate the issue is that you could use paramap to parallelize the combined processing of T1 and T2. Still using the analogy above, each person would paint both the face and the body but both could work at the same time and place back the figurine in the right order on the conveyor belt once they are done. Is that what you were suggesting in the second post?

Another solution at the protocol level would be for the sink to issue as many reads as there are stages in the pipeline before receiving an answer. I think that could be made to work within the current protocol.

Finally, a producer-driven stream does not have the problem because the value is pushed to the next stage as soon as it is ready, no matter how many stages there are. You get parallelism of the stages for free. However, that goes against the design of pull-streams, which are consumer-driven (lazy). From that point of view, it really is a matter of design.

Just to be clear, I am trying to reflect deeply about the protocol because I am comparing it to other stream implementations that use dataflow variables instead of callbacks which could support both producer- and consumer-driven streams. The insight I am trying to clarify is whether the callback protocol @dominictarr designed is a restricted form of dataflow stream that implicitly uses the stack for allocation and deallocation of intermediate values. If that is true, then what is really cool about his design is that although you do not have the full expressive power dataflow streams provide you, you may use much less memory and could be faster than if dataflow variables would be allocated on the heap. Moreover, by using callbacks no special support is required from the runtime other than higher-order functions (function objects).

**Comment by elavoie** [comment]
When a source fails once with an error, should all following reads return the same error?

**Comment by dominictarr** [comment]
@elavoie if you used pull-high-watermark HW inbetween T1 and T2 then it would be closer to 1 item per second. say, you set the watermark to 2, as soon as one item has come through T1, HW will pass it on too T2, but also request another one from T1, this way both T1 and T2 will be opperating near capacity.

Out of the box, pull-streams do not provide any built in paralism, but where needed that should be dropped in by modules that provide the local behaviour required. Such as pull-paramap or pull-high-watermark.

I've generally found pull-paramap to be much more useful than pull-high-watermark,
inpractice, this has been more challenging at the sink end, for that I have
* https://github.com/pull-stream/pull-write
* and recently: https://github.com/dominictarr/async-write ([easy to wrap into a pull-stream]( https://github.com/ssbc/secure-scuttlebutt/blob/validate2/index.js#L140-L155))

I'm interested in adaptive models for this, I think I still have a bit to learn in this area.

**Comment by dominictarr** [comment]
@elavoie your figurine example is good. pull-paramap does preserve order, so if some figurines are painted quickly, but a slow one was issued first, it will buffer the quick ones until the slow one is ready to perserve the order...

However, there is also an `inOrder` option (`async, width, inOrder`) and if that is set to `false` the figurines will just go out in the order they are processed.

Anyway, the point is with a few modules a consumer based stream can do anything a producer based stream can do. On the other hand, you can also add modules to a producer based stream to make it behave more like a consumer based stream, end of the day they are both ways of expressing the same ideas.

**Comment by elavoie** [comment]
@dominictarr it seems paramap can also be used as a bounded buffer if the function it uses does nothing. It will ask at most N values from the source and will wait asking for more values only when the sink has consumed some of them. With N equal to the number of stages in the pipeline you would get the producer behaviour mentioned above.

I moved the specification repository mentioned above to my own user account until I feel it is good enough, it is incomplete at the moment and my current paper is evolving in a way that won't leave me time to complete it soon. So in the meantime, if anyone is interested it is going to be here (https://github.com/elavoie/pull-stream-specification).

I am making progress on a formalization that captures concisely many of the informal rules mentioned above. I will report back once it is complete enough so you guys can give me feedback.

**Comment by dominictarr** [comment]
@elavoie great point! I hadn't thought of that!
